snhst/reduce_hst_data.py

```diff
@@ -17,178 +17,217 @@
+    visit_meta_data = sort_raw_data(raw_images)
+    hst_reference = get_default_reference_hst(visit_meta_data)
+    instrument_string = go_to_visit(hst_reference)
 
     # Make the overall HST template image
-    if reference_image is not None:
-        options['ra'], options['dec'] = get_center_of_image(reference_image)
+    if ground_reference is not None:
+
+        drizzled_template_filename, drizzled_template_catalog = match_and_drizzle('match_template', instrument_string, options)
+        drizzled_template_header = fits.getheader(drizzled_template_filename, extname='SCI')
+
+        # Get the science hdu of the ground reference (may be compressed)
+        reference_hdulist = fits.open(os.path.join(topdir, ground_reference))
+        for reference_hdu in reference_hdulist:
+            if reference_hdu.header.get('NAXIS'):
+                break
+
+        if not os.path.exists('ground_reference.cat'):
+            # Reproject ground image onto drizzled HST frame
+            data, footprint = reproject.reproject_interp(reference_hdu, drizzled_template_header)
+            # Run sep on the ground image to make the ground catalog
+            make_sep_catalog(data, options['sep'], 'match_template/ground_reference.cat')
+
+        # Calculate the offset between the frames
+        # (both catalogs are projected to the WCS of the drizzled template)
+        offsets = wcs.calculate_wcs_offset(drizzled_template_header, drizzled_template_catalog, 'match_template/ground_reference.cat', max_offset=2.)
+    else:
+        offsets = None
+
+    hst_reference_filename, hst_reference_catalog = match_and_drizzle('hst_template', instrument_string, options, offsets)
+
+    # drizzle each visit separately
+    for visit in visit_meta_data:
+        if (visit == hst_reference).all():
+            continue
+        instrument_string = go_to_visit(visit)
+        visit_template_filename, visit_template_catalog = match_and_drizzle('match_template', instrument_string, options)
+        visit_template_header = fits.getheader(visit_template_filename, extname='SCI')
+        offsets = wcs.calculate_wcs_offset(visit_template_header, visit_template_catalog, hst_reference_catalog, max_offset=2.)
+        match_and_drizzle('hst_template', instrument_string, options, offsets)
+
+    # drizzle all visits together
+    output_images = visit_meta_data.group_by(['instrument', 'detector', 'subarray', 'filter'])
+    for output_image in output_images.groups.keys:
+        instrument_string = go_to_visit(output_image)
+        for image in get_raw_image_filenames('visit*'):
+            fits_utils.copy_if_not_exists(image, '.')
+        filter_template_filename, filter_template_catalog = match_and_drizzle('match_template', instrument_string, options)
+        filter_template_header = fits.getheader(filter_template_filename, extname='SCI')
+        offsets = wcs.calculate_wcs_offset(filter_template_header, filter_template_catalog, hst_reference_catalog, max_offset=2.)
```

Comment:
I haven't addressed the first two comments. We still offset by visit+filter instead of only by visit. But we do avoid recalculating the offsets later in the process (third comment).